### PR TITLE
[JAX] add bias to tri-mul

### PR DESCRIPTION
**Pull request by mariogeiger**

### CODE REVIEW

File: cuequivariance_jax/cuequivariance_jax/triangle/_sigmoid_gated_dual_gemm.py [hunk 1]
```diff
@@ -76,25 +108,59 @@ def _abstract_eval_fwd(
...
+    b1: jax.Array,
+    b2: jax.Array,
+    mask: jax.Array,
+    *,
+    two_inputs: bool,
+    transpose_out: bool,
+    precision: Precision,
+    has_b1: bool,
+    has_b2: bool,
+    has_mask: bool,
```
Comment (hsadasivan) [inline cuequivariance_jax/cuequivariance_jax/triangle/_sigmoid_gated_dual_gemm.py:154]: Why do we expose has_mask in API? I

File: cuequivariance_jax/cuequivariance_jax/triangle/_sigmoid_gated_dual_gemm.py [hunk 2]
```diff
@@ -115,20 +188,25 @@ def _reference_forward(x1, x2, w1, w2, mask, two_inputs, transpose_out, precisio
 
 
 def _triton_forward(
```
Comment (hsadasivan) [inline cuequivariance_jax/cuequivariance_jax/triangle/_sigmoid_gated_dual_gemm.py:190]: need to set precision arg correctly like in torch somewhere inside this before calling triton kernel

File: cuequivariance_jax/cuequivariance_jax/triangle/_sigmoid_gated_dual_gemm.py (hunk 1 above)
Comment (hsadasivan) [reply cuequivariance_jax/cuequivariance_jax/triangle/_sigmoid_gated_dual_gemm.py:154]: >sigmoid_gated_dual_gemm is the function that is exposed
Ah, got it. My bad. Thanks.

File: cuequivariance_jax/tests/triangle_multiplicative_update_test.py [hunk 3]
```diff
@@ -108,7 +105,7 @@ def test_compare_with_pytorch(direction, use_mask):
             mask=mask_torch,
             **weights_torch,
             eps=eps,
-            precision=PrecisionTorch.DEFAULT,
+            precision="IEEE",
         )
 
     out_jax = triangle_multiplicative_update_jax(
```
Comment (hsadasivan) [inline cuequivariance_jax/tests/triangle_multiplicative_update_test.py:111]: @mariogeiger 
Should we add test for new features like trimul with bias here from torch?

File: cuequivariance_jax/cuequivariance_jax/triangle/__init__.py [hunk 4]
```diff
@@ -12,22 +12,14 @@
 # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 # See the License for the specific language governing permissions and
 # limitations under the License.
-from ._layer_norm_transpose import layer_norm_transpose
-from ._sigmoid_gated_dual_gemm import (
-    Precision,
-    sigmoid_gated_dual_gemm,
-    sigmoid_gated_dual_gemm_dual_x,
-)
+from ._sigmoid_gated_dual_gemm import Precision
```
Comment (hsadasivan) [inline cuequivariance_jax/cuequivariance_jax/triangle/__init__.py:15]: Should probably move precision to a utils.py outside gated gemm

Comment (mariogeiger) [reply cuequivariance_jax/cuequivariance_jax/triangle/__init__.py]: done

File: cuequivariance_jax/cuequivariance_jax/triangle/_sigmoid_gated_dual_gemm.py [hunk 5]
```diff
@@ -220,18 +309,17 @@ def _triton_backward(
         jax.ShapeDtypeStruct(shape=(M, N), dtype=x1.dtype),  # grad_xw1
         jax.ShapeDtypeStruct(shape=(M, N), dtype=x1.dtype),  # grad_xw2
     ]
-    if mask is not None:
-        num_tiles_n = triton.cdiv(N, TILE_N)
-        out_shapes.append(
-            jax.ShapeDtypeStruct(shape=(num_tiles_n, M), dtype=mask.dtype)
-        )
+    num_tiles_n = triton.cdiv(N, TILE_N)
```
Comment (hsadasivan) [inline cuequivariance_jax/cuequivariance_jax/triangle/_sigmoid_gated_dual_gemm.py:287]: why did we remove the `mask is not None` check

File: cuequivariance_jax/cuequivariance_jax/triangle/_sigmoid_gated_dual_gemm.py [hunk 6]
```diff
@@ -362,47 +440,50 @@ def fn(x):
...
+            "M": m,
+            "N": n,
+            "K": 128,
+            "dtype_input": dt,
+            "two_inputs": ti,
+            "precision": p,
+        }
+        for n in (128, 256)
+        for ti in (True, False)
+        for m in range(32, 2048, 32)
```
Comment (hsadasivan) [inline cuequivariance_jax/cuequivariance_jax/triangle/_sigmoid_gated_dual_gemm.py:459]: why is there a difference in` m` range compared to torch?

File: cuequivariance_jax/cuequivariance_jax/triangle/_sigmoid_gated_dual_gemm.py [hunk 7]
```diff
@@ -497,107 +556,194 @@ def forward_fn(x1, x2, w1, w2, mask):
...
+    w1: jax.Array,
+    w2: jax.Array,
+    b1: jax.Array,
+    b2: jax.Array,
+    mask: jax.Array,
+    two_inputs: bool,
+    transpose_out: bool,
+    precision: Precision,
+    fallback: bool = False,
+    has_b1: bool = True,
```
Comment (hsadasivan) [inline cuequivariance_jax/cuequivariance_jax/triangle/_sigmoid_gated_dual_gemm.py:581]: Maybe set bias defaults to False?

File: cuequivariance_jax/tests/triangle_multiplicative_update_test.py [hunk 8]
```diff
@@ -57,16 +67,14 @@ def create_weights(hidden_dim, seed=42, device=None):
 @pytest.mark.slow
 @pytest.mark.parametrize("direction", ["outgoing", "incoming"])
 @pytest.mark.parametrize("use_mask", [False, True])
-def test_compare_with_pytorch(direction, use_mask):
```
Comment (hsadasivan) [inline cuequivariance_jax/tests/triangle_multiplicative_update_test.py:60]: Also, what about adding the jax version of the test: `test_all_biases_fallback_to_torch`. This might be super important as you have JAX reference to switch to.

File: cuequivariance_jax/cuequivariance_jax/triangle/_triangle_multiplicative_update.py [hunk 9]
```diff
@@ -325,7 +400,14 @@ def triangle_multiplicative_update(
...
-        x_in, x_out, g_out_weight, p_out_weight, precision=precision, fallback=fallback
+        x_in,
+        x_out,
+        g_out_weight,
+        p_out_weight,
+        b1=g_out_bias,
+        b2=p_out_bias,
+        precision=precision,
+        fallback=fallback,
     )
```
Comment (hsadasivan) [inline cuequivariance_jax/cuequivariance_jax/triangle/_triangle_multiplicative_update.py:412]: Another one unrelated to PR, I think we should change 
`if hidden_dim % 64 != 0:` to check for multiples of 32.

File: cuequivariance_jax/cuequivariance_jax/triangle/_sigmoid_gated_dual_gemm.py (hunk 6 above)
Comment (mariogeiger) [reply cuequivariance_jax/cuequivariance_jax/triangle/_sigmoid_gated_dual_gemm.py]: good catch, changed to `for m in range(32, 2048 * 2048 + 1, 32)`

File: cuequivariance_jax/cuequivariance_jax/triangle/_triangle_multiplicative_update.py (hunk 9 above)
Comment (mariogeiger) [reply cuequivariance_jax/cuequivariance_jax/triangle/_triangle_multiplicative_update.py:414]: changed

**Review by hsadasivan — APPROVED** [review]
